From 98d59d08aae62fc237ea72f0b6d295d1c23a8a21 Mon Sep 17 00:00:00 2001
Message-ID: <98d59d08aae62fc237ea72f0b6d295d1c23a8a21.1771966444.git.lorenzo@kernel.org>
In-Reply-To: <2cd15ef3379025ea1ae1f6c36c70621aa578c270.1771966444.git.lorenzo@kernel.org>
References: <2cd15ef3379025ea1ae1f6c36c70621aa578c270.1771966444.git.lorenzo@kernel.org>
From: Lorenzo Bianconi <lorenzo@kernel.org>
Date: Sat, 27 Dec 2025 11:54:31 +0100
Subject: [bpf-next 13/16] net: airoha: Move GDM forward port configuration in
 ndo_open/ndo_stop callbacks

This change allows to set GDM forward port configuration to
FE_PSE_PORT_DROP stopping the network device.
Moreover, PPE firmware requires to use PPE1 for GDM3 or GDM4.

Signed-off-by: Lorenzo Bianconi <lorenzo@kernel.org>
---
 drivers/net/ethernet/airoha/airoha_eth.c | 22 +++++++++++++---------
 1 file changed, 13 insertions(+), 9 deletions(-)

--- a/drivers/net/ethernet/airoha/airoha_eth.c
+++ b/drivers/net/ethernet/airoha/airoha_eth.c
@@ -1629,6 +1629,7 @@ static int airoha_dev_open(struct net_de
 	struct airoha_gdm_dev *dev = netdev_priv(netdev);
 	struct airoha_gdm_port *port = dev->port;
 	struct airoha_qdma *qdma = dev->qdma;
+	u32 pse_port = FE_PSE_PORT_PPE1;
 
 #if defined(CONFIG_PCS_AIROHA)
 	if (airhoa_is_phy_external(port)) {
@@ -1666,12 +1667,21 @@ static int airoha_dev_open(struct net_de
 			GLOBAL_CFG_RX_DMA_EN_MASK);
 	atomic_inc(&qdma->users);
 
+	if (port->id == AIROHA_GDM2_IDX &&
+	    airoha_ppe_is_enabled(qdma->eth, 1)) {
+		/* For PPE2 always use secondary cpu port. */
+		pse_port = FE_PSE_PORT_PPE2;
+	}
+	airoha_set_gdm_port_fwd_cfg(qdma->eth, REG_GDM_FWD_CFG(port->id),
+				    pse_port);
+
 	return 0;
 }
 
 static int airoha_dev_stop(struct net_device *netdev)
 {
 	struct airoha_gdm_dev *dev = netdev_priv(netdev);
+	struct airoha_gdm_port *port = dev->port;
 	struct airoha_qdma *qdma = dev->qdma;
 	int i, err;
 
@@ -1683,6 +1693,9 @@ static int airoha_dev_stop(struct net_de
 	for (i = 0; i < ARRAY_SIZE(qdma->q_tx); i++)
 		netdev_tx_reset_subqueue(netdev, i);
 
+	airoha_set_gdm_port_fwd_cfg(qdma->eth, REG_GDM_FWD_CFG(port->id),
+				    FE_PSE_PORT_DROP);
+
 	if (atomic_dec_and_test(&qdma->users)) {
 		airoha_qdma_clear(qdma, REG_QDMA_GLOBAL_CFG,
 				  GLOBAL_CFG_TX_DMA_EN_MASK |
@@ -1783,7 +1796,6 @@ static int airoha_dev_init(struct net_de
 	struct airoha_gdm_port *port = dev->port;
 	struct airoha_qdma *qdma = dev->qdma;
 	struct airoha_eth *eth = qdma->eth;
-	u32 pse_port = FE_PSE_PORT_PPE1;
 	int i;
 
 	airoha_set_macaddr(dev, netdev->dev_addr);
@@ -1799,19 +1811,11 @@ static int airoha_dev_init(struct net_de
 			if (err)
 				return err;
 		}
-		fallthrough;
-	case AIROHA_GDM2_IDX:
-		if (airoha_ppe_is_enabled(eth, 1)) {
-			/* For PPE2 always use secondary cpu port. */
-			pse_port = FE_PSE_PORT_PPE2;
-			break;
-		}
 		break;
 	default:
 		break;
 	}
 
-	airoha_set_gdm_port_fwd_cfg(eth, REG_GDM_FWD_CFG(port->id), pse_port);
 	for (i = 0; i < eth->soc->num_ppe; i++)
 		airoha_ppe_set_def_cport(dev, i);
 
